src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -12197,75 +12067,42 @@ private InitializerExpressionSyntax ParseObjectOrCollectionInitializer()
+                static @this => @this.IsInitializerMember(),
+                static @this => @this.ParseObjectOrCollectionInitializerMember(),
+                static (@this, openBrace, list, expectedKind, _) => @this.SkipBadInitializerListTokens(openBrace, list, expectedKind),
+                allowTrailingSeparator: true);
+
+            var kind = isObjectInitializer(initializers) ? SyntaxKind.ObjectInitializerExpression : SyntaxKind.CollectionInitializerExpression;
 
             return _syntaxFactory.InitializerExpression(
                 kind,
                 openBrace,
-                _pool.ToListAndFree(initializers),
+                initializers,
                 this.EatToken(SyntaxKind.CloseBraceToken));
-        }
-
-        private void ParseObjectOrCollectionInitializerMembers(
-            ref SyntaxToken startToken,
-            SeparatedSyntaxListBuilder<ExpressionSyntax> list,
-            out SyntaxKind kind)
-        {
-            // Empty initializer list must be parsed as an object initializer.
-            kind = SyntaxKind.ObjectInitializerExpression;
 
-            if (this.CurrentToken.Kind != SyntaxKind.CloseBraceToken)
+            static bool isObjectInitializer(SeparatedSyntaxList<ExpressionSyntax> initializers)
             {
-tryAgain:
-                if (this.IsInitializerMember() || this.CurrentToken.Kind == SyntaxKind.CommaToken)
-                {
-                    // We have at least one initializer expression.
-                    // If at least one initializer expression is a named assignment, this is an object initializer.
-                    // Otherwise, this is a collection initializer.
-                    kind = SyntaxKind.CollectionInitializerExpression;
-
-                    // first argument
-                    list.Add(this.ParseObjectOrCollectionInitializerMember(ref kind));
-
-                    // additional arguments
-                    int lastTokenPosition = -1;
-                    while (IsMakingProgress(ref lastTokenPosition))
-                    {
-                        if (this.CurrentToken.Kind == SyntaxKind.CloseBraceToken)
-                        {
-                            break;
-                        }
-                        else if (this.CurrentToken.Kind == SyntaxKind.CommaToken || this.IsInitializerMember())
-                        {
-                            list.AddSeparator(this.EatToken(SyntaxKind.CommaToken));
-
-                            // check for exit case after legal trailing comma
-                            if (this.CurrentToken.Kind == SyntaxKind.CloseBraceToken)
-                            {
-                                break;
-                            }
+                // Empty initializer list must be parsed as an object initializer.
+                if (initializers.Count == 0)
+                    return true;
 
-                            list.Add(this.ParseObjectOrCollectionInitializerMember(ref kind));
```

Comment:
this old logic deviated from every other function that parsed separated syntax lists, but allows trailing commas.  This one would unilaterally then try to parse the next item, even if we weren't on the start of a possible item.  That would lead to multiple diagnostics instead of just one.   I figured deviating here was acceptable.  It does change 3 tests, but tehy all reduce in the number of diags reported.